File: src/test/OLD_TEST_MATLAB/testEMGLLF.m
function[] = testEMGLLF()

	testFolder = 'data/';
	delimiter = '\n';

	%get dimensions
	dimensions = dlmread(strcat(testFolder,'dimensions'), delimiter);
	n = dimensions(1);
	p = dimensions(2);
	m = dimensions(3);
	k = dimensions(4);

	%get all input arrays
	phiInit = reshape(dlmread(strcat(testFolder,'phiInit'), delimiter), p, m, k);
	rhoInit = reshape(dlmread(strcat(testFolder,'rhoInit'), delimiter), m, m, k);
	piInit = transpose(dlmread(strcat(testFolder,'piInit'), delimiter));
	gamInit = reshape(dlmread(strcat(testFolder,'gamInit'), delimiter), n, k);
	mini = int64(dlmread(strcat(testFolder,'mini'), delimiter));
	maxi = int64(dlmread(strcat(testFolder,'maxi'), delimiter));
	gamma = dlmread(strcat(testFolder,'gamma'), delimiter);
	lambda = dlmread(strcat(testFolder,'lambda'), delimiter);
	X = reshape(dlmread(strcat(testFolder,'X'), delimiter), n, p);
	Y = reshape(dlmread(strcat(testFolder,'Y'), delimiter), n, m);
	tau = dlmread(strcat(testFolder,'tau'), delimiter);

	%run EMGLLF.m
	[phi,rho,pi,LLF,S] = EMGLLF(phiInit,rhoInit,piInit,gamInit,mini,maxi,gamma,lambda,X,Y,tau);

	%get all stored outputs
	ref_phi = reshape(dlmread(strcat(testFolder,'phi'), delimiter), p, m, k);
	ref_rho = reshape(dlmread(strcat(testFolder,'rho'), delimiter), m, m, k);
	ref_pi = dlmread(strcat(testFolder,'pi'), delimiter);
	ref_LLF = dlmread(strcat(testFolder,'LLF'), delimiter);
	ref_S = reshape(dlmread(strcat(testFolder,'S'), delimiter), p, m, k);

	%check that output correspond to stored output
	tol = 1e-5;
	checkOutput('phi',phi,ref_phi,tol);
	checkOutput('rho',rho,ref_rho,tol);
	checkOutput('pi',pi,ref_pi,tol);
	checkOutput('LLF',LLF,ref_LLF,tol);
	checkOutput('S',S,ref_S,tol);

end

File: src/test/test.EMGLLF.c
#include "EMGLLF.h"
#include "utils.h"
#include <stdlib.h>

int main(int argc, char** argv)
{
	////////////
	// INPUTS //
	////////////

	int* dimensions = readArray_int("dimensions");
	int n = dimensions[0];
	int p = dimensions[1];
	int m = dimensions[2];
	int k = dimensions[3];
	free(dimensions);

	float* phiInit = readArray_real("phiInit");
	float* rhoInit = readArray_real("rhoInit");
	float* piInit = readArray_real("piInit");
	float* gamInit = readArray_real("gamInit");
	int mini = read_int("mini");
	int maxi = read_int("maxi");
	float gamma = read_real("gamma");
	float lambda = read_real("lambda");
	float* X = readArray_real("X");
	float* Y = readArray_real("Y");
	float tau = read_real("tau");

	/////////////
	// OUTPUTS //
	/////////////

	float* phi = (float*)malloc(p*m*k*sizeof(float));
	float* rho = (float*)malloc(m*m*k*sizeof(float));
	float* pi = (float*)malloc(k*sizeof(float));
	float* LLF = (float*)malloc(maxi*sizeof(float));
	float* S = (float*)malloc(p*m*k*sizeof(float));

	////////////////////
	// Call to EMGLLF //
	////////////////////

	EMGLLF_core(phiInit,rhoInit,piInit,gamInit,mini,maxi,gamma,lambda,X,Y,tau,
		phi,rho,pi,LLF,S,
		n,p,m,k);

	free(phiInit);
	free(rhoInit);
	free(piInit);
	free(gamInit);
	free(X);
	free(Y);

	// Compare to reference outputs
	float* ref_phi = readArray_real("phi");
	compareArray_real("phi", phi, ref_phi, p*m*k);
	free(phi);
	free(ref_phi);

	float* ref_rho = readArray_real("rho");
	compareArray_real("rho", rho, ref_rho, m*m*k);
	free(rho);
	free(ref_rho);

	float* ref_pi = readArray_real("pi");
	compareArray_real("pi", pi, ref_pi, k);
	free(pi);
	free(ref_pi);

	float* ref_LLF = readArray_real("LLF", maxi);
	compareArray_real("LLF", LLF, ref_LLF);
	free(LLF);
	free(ref_LLF);

	float* ref_S = readArray_real("S");
	compareArray_real("S", S, ref_S, p*m*k);
	free(S);
	free(ref_S);

	return 0;
}
